src/main/java/com/commercetools/sync/services/impl/ProductServiceImpl.java

```diff
@@ -74,35 +116,103 @@ public ProductServiceImpl(@Nonnull final ProductSyncOptions syncOptions) {
 
   @Nonnull
   @Override
-  public CompletionStage<Set<Product>> fetchMatchingProductsByKeys(
+  public CompletionStage<Set<ProductProjection>> fetchMatchingProductsByKeys(
       @Nonnull final Set<String> productKeys) {
 
-    return fetchMatchingResources(
+    return fetchMatchingResourcesInternal(
         productKeys,
+        resource -> resource.getKey(),
         (keysNotCached) ->
-            ProductQuery.of().withPredicates(buildProductKeysQueryPredicate(keysNotCached)));
+            ProductProjectionQuery.ofStaged()
+                .withPredicates(buildProductKeysQueryPredicate(keysNotCached)));
+  }
+
+  <Q extends SphereRequest<PagedQueryResult<T>>, T extends ProductProjection>
+      CompletionStage<Set<ProductProjection>> fetchMatchingResourcesInternal(
+          @Nonnull final Set<String> keys,
+          @Nonnull final Function<T, String> keyMapper,
+          @Nonnull final Function<Set<String>, Q> keysQueryMapper) {
+
+    if (keys.isEmpty()) {
+      return CompletableFuture.completedFuture(Collections.emptySet());
+    }
+
+    final List<List<String>> chunkedKeys = ChunkUtils.chunk(keys, CHUNK_SIZE);
+
+    List<Q> keysQueryMapperList =
+        chunkedKeys.stream()
+            .map(_keys -> keysQueryMapper.apply(new HashSet<>(_keys)))
```

Comment:
That part is updated after your changes, don't know why the conflicts are resolved without considering the master branch changes. Here: https://github.com/commercetools/commercetools-sync-java/commit/c3e039947bc13f68ab31f317fdeb6fbbff2bd937#diff-d0e88448db799d7fa3365c003712f1839aaa20e2f06b1838c926d2c71b5b2018R274-R277

Reply:
The change was done in the BaseService but I had to replicate the code to the Productservice, because "productprojections" are not compatible with the method in the base service (Productprojection don`t implement the resource interface)

Reply:
simply you need to add to the query, otherwise the query will return 20 product only as default limit is 20.
````
.withLimit(CHUNK_SIZE)
 .withFetchTotal(false)
````